cmd/test-bot.rb

```diff
@@ -83,10 +83,12 @@ def test_bot
 
   def setup_argv_and_env
     jenkins = !ENV["JENKINS_HOME"].nil?
+    self_hosted = !ENV["HOMEBREW_SELF_HOSTED_RUNNER"].nil?
 
     github_actions = !ENV["GITHUB_ACTIONS"].nil?
     if github_actions
-      ARGV << "--verbose" << "--ci-auto" << "--no-pull"
+      ARGV << "--verbose" << "--ci-auto"
+      ARGV << "--no-pull" unless self_hosted
```

Comment:
```suggestion
      ARGV << "--no-pull" unless github_actions_self_hosted
```

but I think it'd be preferable to avoid needing `--no-pull` here if possible.

Reply:
Either:

* `brew test-bot` grabs the correct set of commits with `brew pull`, or
* we duplicate that logic in the workflow.

Re: https://github.com/Homebrew/homebrew-core/pull/50468#discussion_r383032833

This cannot use `actions/checkout` because, like the Jenkins setup, the VMs are persistent. Using `actions/checkout` is only possible on Linux [because we remove `$(brew --repo homebrew/core)`](https://github.com/Homebrew/linuxbrew-core/blob/f20b3f97559466a46cb87cb392201ea90c509b37/.github/workflows/build-bottles.yml#L15-L20) and the runners are ephemeral so it doesn't matter if the tap folder gets taken out.

Reply:
> This cannot use `actions/checkout` because, like the Jenkins setup, the VMs are persistent.

Can you elaborate a bit on this? Would love to get e.g. `brew test-bot` to a point where we can use actions/checkout.

Reply:
I thought I already did, in the next sentence. What do you need clarification about?

Reply:
You could `rm -rf $(brew --repo homebrew/core)` and then run `actions/checkout`.

Reply:
A full clone is something like 2-4 minutes per CI job. I didn’t think that was an acceptable trade off given we avoid it in the dispatch build bottle job on Linux, and that we changed the Dockerfile to include full clones of homebrew/core.

Reply:
Good point. I agree!
In that case I'd suggest
```
cd $(brew --repo homebrew/core) && brew reset --hard $GITHUB_SHA`
```
either in the GitHub Action before running `brew test-bot` or in `brew test-bot`.

Reply:
@jonchang I don't think I understand why persistence means we can't use it. Is it because it can only do a full clone every time and not/never e.g. update an existing clone?

Reply:
Shaun: I tried that already. You need to do a `git fetch origin refs/pull/pr/merge` as the existing repository doesn’t know what object `github.sha` is because there’s no ref that points to it. 

Mike: I didn’t see anything in the documentation that said actions/checkout could do that. It only refers to new clones (shallow or full).

Reply:
Thanks @jonchang!

Reply:
```sh
cd $(brew --repo homebrew/core)
git fetch origin "$GITHUB_REF"
git checkout master
git reset --hard "$GITHUB_SHA"
```
?

Reply:
Yeah, this works: https://github.com/Homebrew/linuxbrew-core/pull/19715

`actions/checkout@v2` to an existing directory fails: https://github.com/Homebrew/linuxbrew-core/runs/468679652

Reply:
Thanks for investigating ❤️